src/scheduler.rs

```diff
@@ -0,0 +1,1056 @@
+//!
+//! A Kubernetes pod scheduler is responsible for assigning pods to eligible nodes. To achieve this,
+//! the scheduler may use different strategies.
+//!
+//! This module provides traits and implementations for a scheduler with memory called [`StickyScheduler`]
+//! and two pod placement strategies : [`ScheduleStrategy::GroupAntiAffinity`] and [`ScheduleStrategy::Hashing`].
+//!
+//! The former strategy means that no two pods belonging to the same role+group pair may be scheduled on the
+//! same node. This is useful for bare metal scenarios.
+//!
+//! The latter strategy hashes pods to nodes without any regards to the node load.
+//!
+//! The scheduler implements the idea of "preferred nodes" where pods should be scheduled.
+//! Whether a preferred node is selected for a pod depends not only of the node's eligibility but also
+//! on the strategy used.
+//!
+//! One implementation for a preferred nodes provider is the [`K8SUnboundedHistory`] that keeps
+//! track of pod placements and reuses the nodes in the future. It uses the K8S resource to store
+//! and retrieve past pod to node assignments. The requirement for this preferred node provider is
+//! that pod id's are "stable" and have a semantic known to the calling operator.
+//!
+//!
+use std::collections::{BTreeMap, HashSet};
+use std::fmt::{Debug, Display, Formatter};
+
+use crate::client::Client;
+use crate::error::OperatorResult;
+use crate::labels;
+use crate::role_utils::EligibleNodesForRoleAndGroup;
+use k8s_openapi::api::core::v1::{Node, Pod};
+use kube::api::Resource;
+use schemars::JsonSchema;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::cell::RefCell;
+use std::collections::btree_map::Iter;
+use std::collections::hash_map::DefaultHasher;
+use std::convert::TryFrom;
+use std::hash::{Hash, Hasher};
+use std::ops::DerefMut;
+use tracing::{error, warn};
+
+const DEFAULT_NODE_NAME: &str = "<no-nodename-set>";
+const SEMICOLON: &str = ";";
```

Comment:
Could be named according to its purpose, like `POD_IDENTITY_SEPARATOR`.

Reply:
Agreed that it will be done in the next round of refactoring.